### PR TITLE
`sticky` LoadBalancer implementation

### LINKED ISSUE #271: sticky load balancer
Select one service instance and use it until it fails. Then switch to a different one.

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
fixes #271

### CODE REVIEW

File: docs/load-balancer/sticky.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Sticky Load Balancing
+
+The sticky load balancer selects a single service instance and keeps using it until it fails.
+Then, it selects the next one.
+
+It is possible to configure a back off time to specify for how long a service instance that failed should not be retried.
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:6]:
```suggestion
It is possible to configure a backoff time to specify for how long a failing service instance should not be retried.
```

File: docs/load-balancer/sticky.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+Precisely, the load balancer works as follows:
+
+* if the previously selected service instance has not failed, and is still available, return it
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:10]:
```suggestion
* if the previously selected service instance has not failed, and is still available, return it;
```

File: docs/load-balancer/sticky.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+* else if exists, return the first available service instance that has no recorded failure
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:11]:
```suggestion
* else if the service instance is still available, but has failed during a previous invocation, return the first available service instance that has no recorded failure;
```

File: docs/load-balancer/sticky.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+* else, find the available instance that has failed the longest time ago, and
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:12]:
```suggestion
* otherwise, find the available instance that was the first to record a failure, and
```

File: docs/load-balancer/sticky.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    * if the back-off time since failure passed, return it
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:13]:
```suggestion
    * if the backoff time since the failure passed, return it;
```

File: docs/load-balancer/sticky.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    * else, report lack of applicable services (throw `NoAcceptableServiceInstanceFoundException`)
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:14]:
```suggestion
    * or, throw an `NoAcceptableServiceInstanceFoundException` as no acceptable instances are available.
```

File: docs/load-balancer/sticky.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+```properties
+stork.my-service.service-discovery=...
+stork.my-service.service-discovery...=...
+stork.my-service.load-balancer.type=sticky
+```
+
+
+The following attributes are supported:
+
+--8<-- "load-balancer/least-response-time/target/classes/META-INF/stork-docs/least-response-time-lb-attributes.txt"
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:29]: That looks wrong...

File: load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancer.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            }
+            failedInstanceIterator.remove();
+            ServiceInstance serviceInstance = serviceInstances.get(instanceId);
+            if (serviceInstance != null) {
+                lastSelected = new ServiceInstanceWithStatGathering(serviceInstance, this);
+                return lastSelected;
+            }
+        }
+        throw new NoAcceptableServiceInstanceFoundException("Each of the available service instances failed " +
+                "within the configured failure-back-off-time");
```
Comment (cescoffier) [inline load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancer.java:88]: I think we should use `backoff`

File: load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancer.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    void recordEndAtTime(long serviceInstanceId, Throwable error, long currentTime) {
+        if (error != null) {
+            failedInstances.put(serviceInstanceId, currentTime);
+            lastSelected = null;
+        }
+    }
+
+    @Deprecated // for tests only
+    LinkedHashMap<Long, Long> getFailedInstances() {
+        return failedInstances;
```
Comment (cescoffier) [inline load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancer.java:106]: to avoid misusage, should we return a copy?

File: load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancerProvider.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+@LoadBalancerType(StickyLoadBalancerProvider.TYPE)
+@LoadBalancerAttribute(name = StickyLoadBalancerProvider.FAILURE_BACK_OFF_TIME, defaultValue = "0", description = "After how much time, "
+        +
+        "a service instance that has failed can be reused")
+public class StickyLoadBalancerProvider
+        implements LoadBalancerProvider<StickyConfiguration> {
+
+    static final String TYPE = "sticky";
+    static final String FAILURE_BACK_OFF_TIME = "failure-back-off-time";
```
Comment (cescoffier) [inline load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancerProvider.java:18]: I think we should use `backoff`

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: docs/load-balancer/sticky.md (hunk 3 above)
Comment (michalszynkiewicz) [reply docs/load-balancer/sticky.md:11]: the corrected sentences is not true. If no service was selected previously, we should fall under this branch too. Also it may happen that the previously selected instance did not fail but is just not availalbe anymore (service discovery-wise)

File: docs/load-balancer/sticky.md (hunk 4 above)
Comment (michalszynkiewicz) [reply docs/load-balancer/sticky.md:12]: that's not technically true. 
Let's assume we have services A and B.
A failed first, then B failed while backoff time had passed for A. So we used A. Then A failed.
At this point we don't select A.

Comment (michalszynkiewicz) [reply docs/load-balancer/sticky.md:12]: I reworded it a bit

File: load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancer.java (hunk 9 above)
Comment (michalszynkiewicz) [reply load-balancer/sticky/src/main/java/io/smallrye/stork/loadbalancer/random/StickyLoadBalancer.java]: done

File: docs/load-balancer/sticky.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# Sticky Load Balancing
+
+The sticky load balancer selects a single service instance and keeps using it until it fails.
+Then, it selects the next one.
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:4]: instead of `next`, I would say `another` one.

File: docs/load-balancer/sticky.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+It is possible to configure a backoff time to specify for how long a failing service instance should not be retried.
+
+Precisely, the load balancer works as follows:
+
+* if the previously selected service instance has not failed, and is still available, return it;
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:10]: Maybe add the initial state: if no instance has been selected, select the first instance from the list.
(actually, random may be good too).

File: docs/load-balancer/sticky.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+* else if exists, return the first available service instance that has no recorded failure;
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:11]: I would reword the if exists: if the instance is still available, but has failed a previous invocation...

File: docs/load-balancer/sticky.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+* else, find the available instance for which the time since the last failure is the greatest, and
```
Comment (cescoffier) [inline docs/load-balancer/sticky.md:12]:
```suggestion
* else, find an available instance for which the time since the last failure is the longest, and
```

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: docs/load-balancer/sticky.md (hunk 14 above)
Comment (michalszynkiewicz) [reply docs/load-balancer/sticky.md:12]: Why `an` and not `the`? We're talking about a specific instance here.

**Review by cescoffier — APPROVED** [review]
